lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/CompositePostingsEnum.java

```diff
@@ -0,0 +1,165 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.search.uhighlight;
+
+import java.io.IOException;
+import java.util.List;
+
+import org.apache.lucene.index.PostingsEnum;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.PriorityQueue;
+
+
+final class CompositePostingsEnum extends PostingsEnum {
+
+  private static final int NO_MORE_POSITIONS = -2;
+  private final BytesRef term;
+  private final int freq;
+  private final PriorityQueue<BoundsCheckingPostingsEnum> queue;
+
+
+  /**
+   * This class is used to ensure we don't over iterate the underlying
+   * postings enum by keeping track of the position relative to the
+   * frequency.
+   * Ideally this would've been an implementation of a PostingsEnum
+   * but it would have to delegate most methods and it seemed easier
+   * to just wrap the tweaked method.
+   */
+  private static final class BoundsCheckingPostingsEnum {
+
+
+    private final PostingsEnum postingsEnum;
+    private final int freq;
+    private int position;
+    private int nextPosition;
+    private int positionInc = 1;
+
+    private int startOffset;
+    private int endOffset;
+
+    BoundsCheckingPostingsEnum(PostingsEnum postingsEnum) throws IOException {
+      this.postingsEnum = postingsEnum;
+      this.freq = postingsEnum.freq();
+      nextPosition = postingsEnum.nextPosition();
+      position = nextPosition;
+      startOffset = postingsEnum.startOffset();
+      endOffset = postingsEnum.endOffset();
+    }
+
+    private boolean hasMorePositions() throws IOException {
+      return positionInc < freq;
+    }
+
+    /**
+     * Returns the next position of the underlying postings enum unless
+     * it cannot iterate further and returns NO_MORE_POSITIONS;
+     * @return
+     * @throws IOException
+     */
+    private int nextPosition() throws IOException {
+      position = nextPosition;
+      startOffset = postingsEnum.startOffset();
+      endOffset = postingsEnum.endOffset();
+      if (hasMorePositions()) {
+        positionInc++;
+        nextPosition = postingsEnum.nextPosition();
+      } else {
+        nextPosition = NO_MORE_POSITIONS;
+      }
+      return position;
+    }
+
+  }
+
+  CompositePostingsEnum(BytesRef term, List<PostingsEnum> postingsEnums) throws IOException {
+    this.term = term;
+    queue = new PriorityQueue<BoundsCheckingPostingsEnum>(postingsEnums.size()) {
+      @Override
+      protected boolean lessThan(BoundsCheckingPostingsEnum a, BoundsCheckingPostingsEnum b) {
+        return a.position < b.position;
```

Comment:
In the event the positions are equal (e.g. two terms a the same position in which the wildcard matches both), we might want to fall-back on startOffset then endOffset?  Or maybe simply ignore position altogether and just do offsets, so then you needn't even track the position?